universe/auto_syncer.go

```diff
@@ -219,55 +233,173 @@ func (f *FederationEnvoy) syncServerState(ctx context.Context,
+	return nil
+}
+
+// pushProofToFederation attempts to push out a new proof to the current
+// federation in parallel.
+func (f *FederationEnvoy) pushProofToFederation(ctx context.Context,
+	uniID Identifier, key LeafKey, leaf *Leaf, fedServers []ServerAddr,
+	logProofSync bool) {
+
+	log.Infof("Pushing proof to %v federation members, proof_key=%v",
+		len(fedServers), spew.Sdump(key))
 
 	// To push a new proof out, we'll attempt to dial to the remote
 	// registrar, then will attempt to push the new proof directly to the
 	// register.
 	pushNewProof := func(ctx context.Context, addr ServerAddr) error {
-		remoteUniverseServer, err := f.cfg.NewRemoteRegistrar(addr)
-		if err != nil {
-			log.Warnf("cannot push proof unable to connect "+
-				"to remote server(%v): %v", addr.HostStr(),
-				err)
+		// If we are logging proof sync attempts, we will use the
+		// logged version of the push function.
+		if logProofSync {
+			err := f.pushProofToServerLogged(
+				ctx, uniID, key, leaf, addr,
+			)
+			if err != nil {
+				log.Warnf("Cannot push proof via logged "+
+					"server push: %v", err)
+			}
+
 			return nil
 		}
 
-		_, err = remoteUniverseServer.UpsertProofLeaf(
-			ctx, uniID, key, leaf,
-		)
+		// If we are not logging proof sync attempts, we will use the
+		// non-logged version of the push function.
+		err := f.pushProofToServer(ctx, uniID, key, leaf, addr)
 		if err != nil {
-			log.Warnf("cannot push proof to remote "+
-				"server(%v): %v", addr.HostStr(), err)
+			log.Warnf("Cannot push proof: %v", err)
 		}
+
 		return nil
 	}
 
 	// To conclude, we'll attempt to push the new proof to all the universe
 	// servers in parallel.
-	err = fn.ParSlice(ctx, fedServers, pushNewProof)
+	err := fn.ParSlice(ctx, fedServers, pushNewProof)
 	if err != nil {
 		// TODO(roasbeef): retry in the background until successful?
 		log.Errorf("unable to push proof to federation: %v", err)
 		return
 	}
 }
 
+// filterProofSyncPending filters out servers that have already been synced
+// with for the given leaf.
+func (f *FederationEnvoy) filterProofSyncPending(fedServers []ServerAddr,
+	uniID Identifier, key LeafKey) ([]ServerAddr, error) {
+
+	// If there are no servers to filter, then we'll return early. This
+	// saves from querying the database unnecessarily.
+	if len(fedServers) == 0 {
+		return nil, nil
+	}
+
+	ctx, cancel := f.WithCtxQuit()
+	defer cancel()
+
+	// Select all sync push complete log entries for the given universe
+	// leaf. If there are any servers which are sync complete within this
+	// log set, we will filter them out of our target server set.
+	logs, err := f.cfg.FederationDB.QueryFederationProofSyncLog(
+		ctx, uniID, key, SyncDirectionPush,
+		ProofSyncStatusComplete,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to query federation sync log: %w",
+			err)
+	}
+
+	// Filter out servers that we've already pushed to.
+	filteredFedServers := fn.Filter(fedServers, func(a ServerAddr) bool {
+		for idx := range logs {
```

Comment:
nit: could be a lot of looping here if we have a lot of servers (or log entries). IIUC we could iterate over the `logs` once to build a `Set` of host strings, and then check for membership in that.

Not a required change though.

Reply:
I've made this change in the latest commits.